[BEAM-4291] Implements distributed artifact retrieval

jkff opened the pull request:
Taking over @axelmagn 's #5584.

Implements an ArtifactRetrievalService that uses the artifact layout produced by BeamFileSystemArtifactStagingService.

Additionally, fixes a bug where FileIO.ReadableFile would not close the stream, and renames the function that confusingly did not indicate that the stream is not closed.

Tested by adding tests of retrieval service to tests of the staging service.

R: @angoenka

angoenka commented on runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/artifact/BeamFileSystemArtifactRetrievalService.java:
```diff
@@ -0,0 +1,200 @@
+      try (InputStream stream = Channels.newInputStream(FileSystems.open(artifactResourceId))) {
+        int len;
+        while ((len = stream.read(data)) != -1) {
+          hasher.putBytes(data, 0, len);
```
We can avoid md5 computation if `metadata.getMd5() == null`

jkff replied:
Yeah but I'm not sure it's worth adding the redundant if statement - I think in practice MD5 ought to be present (if it's not, we should treat that as a client problem rather than opportunity for optimization).

angoenka commented on runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/artifact/BeamFileSystemArtifactRetrievalService.java:
```diff
@@ -0,0 +1,200 @@
+        ByteString expected = ByteString.copyFrom(Base64.decodeBase64(metadata.getMd5()));
+        ByteString actual = ByteString.copyFrom(hasher.hash().asBytes());
+        if (!actual.equals(expected)) {
+          throw new IllegalStateException(
```
Should we call `responseObserver.onError`

jkff replied:
It gets called below in case of any exception.

jkff commented:
Thanks Robert. I deleted ArtifactSource in this PR now (it did not participate in any live code paths), and added MD5 verification during staging. I want to keep the verification during retrieval as well - the scenario I have in mind is detecting if one job overwrites another job's staged files in the same directory, or if a staged file is incorrectly cached in some way, etc. - it's cheap but can help avoid a number of plausible bugs that otherwise would give a lot of trouble.
